design/one-pager-change-logs.md

```diff
@@ -0,0 +1,410 @@
+which calls into each provider to do specific CUD (Create, Update, and Delete)
+operations on external resources.  Therefore, the managed reconciler is the
+ideal place to generate the change log entries very close to the source of
+truth.
+
+The managed reconciler calls the provider's [`ExternalClient.Observe()`
+method](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L914)
+before any CUD operation is performed, in order to understand the current state
+of the external system before the provider acts upon it. This up to date
+external state will be used to populate the "before" field in the change log
+entry.
+
+If a CUD operation is performed later on in the same `Reconcile()`, then we
+will perform an additional call to this same `Observe()` method directly after
+the CUD operation is performed. This second `Observe()` result will be used to
+populate the "after" field in the change log entry.
+
+Some helpful pointers to where CUD operations are performed in the managed
+reconciler:
+
+* [Create](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L1058)
+* [Update](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L1189)
+* [Delete](https://github.com/crossplane/crossplane-runtime/blob/release-1.16/pkg/reconciler/managed/reconciler.go#L954)
+
+The code paths through the managed reconciler's `Reconcile()` method are already
+a bit complex. One design choice that has helped minimize the complexity and
+assist in readability is that the reconciler returns as early as it can after
+performing an operation. Since we will now be adding an additional `Observe()`
+call after each CUD operation, this will increase the complexity further, but
+is likely unavoidable for us to obtain an immediate understanding of the "after"
+state of a change, without having to wait for another reconcile.
+
+### Storage and Durability
+
+We consider change logs entries to be related to both security and reliability
+of the control plane, and therefore we will have a fairly low tolerance for data
+loss of these logs. Change logs entries will be written to a local location
+immediately, so that a healthy network connection is not required at the time an
+entry is generated. This local location doesn't necessarily need to be a highly
+durable and persistent location itself, the key is that we don't lose any data
+if we temporarily lose network connectivity while entries are being generated.
+
+We should also minimize any effort into building a novel or specific system for
+storing the change logs, so we will favor reusing built in functionality of the
+control plane when possible. For example, we do not want to build any logic to
+perform log rotation, compression, or other log management tasks.
+
+Change log entries will be written to `stdout` so they can be captured by the
```

Comment:
I think we need to have a way for folks to configure their output. I think these logs in large environments will become unruly and super noisy where we may want to have folks say "write them to a seperate file on the node rather than stdout" because my fear here is I want to enable them but I may want them to go through a separate pathway for cost reasons and stored separately or for compliance reasons, retention on these logs may be different.

I get that can be done through a log collector like Otel, Fluentbit, etc; but I think that we will want to have a way to distinguish where the logs be sent out.

Maybe this could be done with a potential additional sidecar like fluentbit that just grabs the logs and then allow folks in a `deploymentruntimeconfig` to specify their fluentbit config for just those sets of logs?

Reply:
>  I think these logs in large environments will become unruly and super noisy where we may want to have folks say "write them to a seperate file on the node rather than stdout" 

I'd explicitly like to avoid this. It starts to imply configuring log compression, log rotation, etc. I'd prefer to let another tool handle that.

> because my fear here is I want to enable them but I may want them to go through a separate pathway for cost reasons and stored separately or for compliance reasons, retention on these logs may be different.

This is why I prefer writing them to stdout from a sidecar container, not the "main" provider process. This way you can configure your log processor of choice (fluent bit etc) to handle them however you like. The way you distinguish them is "logs coming from the change-logs sidecar container".

Reply:
https://github.com/crossplane/crossplane/pull/5822#discussion_r1700092670 could be related here.